### ISSUE TITLE
Suggested Enhancement: Option: Generate Patch

### ISSUE THREAD

**Issue by DavidHaslam** [pnedev/comparePlus#188]
[WinMerge](http://winmerge.org) has a **Tools** option to **Generate Patch**. 

The output file is in standard [diff](https://en.wikipedia.org/wiki/Diff) format that can be used as a [software] patch.

It would be a useful enhancement for the Compare plugin to be able to do likewise.

**Comment by pnedev** [comment]
Great idea, I'll consider implementing it sometime in the future but I can't promise anything as I'm not sure I'll have the time.

**Comment by DavidHaslam** [comment]
The requested feature should not depend upon either **Git** or **SVN** data being present.
cf. Neither are needed by the **WinMerge** tool.

**Comment by b-morgan** [comment]
I'll add my vote for this feature.

**Comment by byzod** [comment]
Here's another vote :D

**Comment by BdR76** [comment]
And here's another vote 👍

I kind of assumed this functionality was already available, because of the SVN/GIT Diff menu items. But I've tried them now and I don't know what they are supposed to do, it just gives a dialog "No Git data found". 🤔 The [Help.md](https://github.com/pnedev/comparePlus/blob/master/Help.md) also doesn't mention it.

**Comment by pnedev** [comment]
@BdR76 ,

Git/SVN compares are used to compare a changed file that is added in Git/SVN repository against its copy in the version control repo.

The counter for patch generation functionality request is incremented :)

**Comment by RafalKisiel** [comment]
Hi,
may I vote for enchancement scope increase? I would see an result export option with boolean operation to choose (\, and, or, xor). I have now a case in which I need to get a intersected part of files.

**Comment by pnedev** [comment]
@RafalKisiel ,
I didn't get your suggestion, sorry.
Could you elaborate?

BR

**Comment by RafalKisiel** [comment]
The base functionality of patch generation (as I understand it) is to export / create new file containig only resulting differences - in boolean operation on collections it is substraction one of this form: _A \(not) B_, _B \(not)A_ or _A xor B_.
There is another boolean operation that could be performed on text files. When you want to get common part of two texts - intersection - this is my case.
My request is to give option to specify what result user want to export.

![Image](https://github.com/user-attachments/assets/d5d2a3ad-f1b3-4f58-8ffd-cec511bc57a9)

Is it more clear now?

**Comment by BdR76** [comment]
I agree with @b-morgan although I do think an export feature, like described in the original request, is still useful.

It's not clear to me what @RafalKisiel 's actual use-case is for this, i.e. to have ability to export the text differences in every and/or combination.

I suspect it might be an edge-case practical application, that would also be feasible by just manipulating the "normal" git diff file.

**Comment by pnedev** [comment]
I'll keep the optional request here as it complements the initial one (those will not necessarily be implemented together).
Thank you for the suggestions guys. 👍

**Comment by pryrt** [comment]
Since it hadn't been specifically mentioned yet in this discussion -- if ComparePlus is going to be able to _generate_ patch files, it should also be able to _apply_ patch files.  (See, for example, [this N++ Community discussion](https://community.notepad-plus-plus.org/topic/26971).)  Since @pnedev allowed the "complement" request to be included as part of this Issue, it also makes sense to have the "inverse" request as part of this Issue, too. ;-)

**Comment by pnedev** [comment]
Patch generation/application is implemented in latest ComparePlus release **[2.0.0](https://github.com/pnedev/comparePlus/releases/tag/cp_2.0.0)**
The patch format is the unified type BUT if compare is run with ignored options enabled the context will be omitted because it will make the patch unable to be applied in some cases.

@RafalKisiel ,
The patch is of standard format so the logical unions/exclusions you suggested cannot be implemented. For those you could use the Diffs Filtering dialog and then copy the visible (unfiltered) lines and save them.

BR
